### PR TITLE
Feature/final code 22 jan 2020

### PR DESCRIPTION
**Pull request by MoteyMousam**
Neutrino settings file updated to support SKLearn BDT PFO characterisation.

### CODE REVIEW

File: settings/PandoraSettings_Master_DUNEFD.xml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <pandora>
     <!-- GLOBAL SETTINGS -->
     <IsMonitoringEnabled>true</IsMonitoringEnabled>
-    <ShouldDisplayAlgorithmInfo>false</ShouldDisplayAlgorithmInfo>
+    <ShouldDisplayAlgorithmInfo>false</ShouldDisplayAlgorithmInfo> <!-- change -->
```
Comment (AndyChappell) [inline settings/PandoraSettings_Master_DUNEFD.xml]: Remove comment

File: settings/PandoraSettings_Master_DUNEFD.xml [hunk 2]
```diff
@@ -13,11 +13,16 @@
         <FilteredCaloHitListName>CaloHitList2D</FilteredCaloHitListName>
         <CurrentCaloHitListReplacement>CaloHitList2D</CurrentCaloHitListReplacement>
     </algorithm>
+    <!--algorithm type = "LArMCParticleMonitoring">
+        <CaloHitListName>CaloHitList2D</CaloHitListName>
+        <MCParticleListName>Input</MCParticleListName>
+	<MinHitsForDisplay>1</MinHitsForDisplay>
+    </algorithm>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Master_DUNEFD.xml]: Remove commented out block

File: settings/PandoraSettings_Master_DUNEFD.xml [hunk 3]
```diff
@@ -13,11 +13,16 @@
...
     <algorithm type = "LArVisualMonitoring">
         <CaloHitListNames>CaloHitListU CaloHitListV CaloHitListW</CaloHitListNames>
+	<MCParticleListNames>Input</MCParticleListNames>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Master_DUNEFD.xml]: Match indentation to surrounding block

File: settings/PandoraSettings_Master_DUNEFD.xml [hunk 4]
```diff
@@ -37,22 +42,23 @@
         <RecreatedClusterListName>RecreatedClusters</RecreatedClusterListName>
         <RecreatedVertexListName>RecreatedVertices</RecreatedVertexListName>
         <VisualizeOverallRecoStatus>false</VisualizeOverallRecoStatus>
+		<InputMCParticleListName>Input</InputMCParticleListName>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Master_DUNEFD.xml]: Match indentation to surrounding block

File: settings/PandoraSettings_Master_DUNEFD.xml [hunk 5]
```diff
@@ -37,22 +42,23 @@
...
         <MCParticleListName>Input</MCParticleListName>
         <PfoListName>RecreatedPfos</PfoListName>
         <UseTrueNeutrinosOnly>false</UseTrueNeutrinosOnly>
         <PrintAllToScreen>false</PrintAllToScreen>
-        <PrintMatchingToScreen>true</PrintMatchingToScreen>
-        <WriteToTree>false</WriteToTree>
-        <OutputTree>Validation</OutputTree>
-        <OutputFile>Validation.root</OutputFile>
+        <PrintMatchingToScreen>false</PrintMatchingToScreen>
+        <WriteToTree>false</WriteToTree> <!-- change -->
```
Comment (AndyChappell) [inline settings/PandoraSettings_Master_DUNEFD.xml]: Remove comment

File: settings/PandoraSettings_Master_DUNEFD.xml [hunk 6]
```diff
@@ -37,22 +42,23 @@
...
+        <OutputTree>test_mousam</OutputTree>
+        <OutputFile>test_mousam.root</OutputFile>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Master_DUNEFD.xml]: Names here should probably revert to the standard Validation and Validation.root versions.

File: settings/PandoraSettings_Master_DUNEFD.xml [hunk 7]
```diff
@@ -37,22 +42,23 @@
...
     </algorithm>
 
-    <algorithm type = "LArVisualMonitoring">
+    <!--algorithm type = "LArVisualMonitoring">
```
Comment (AndyChappell) [inline settings/PandoraSettings_Master_DUNEFD.xml]: Reset to uncommented

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 <pandora>
     <!-- GLOBAL SETTINGS -->
     <IsMonitoringEnabled>true</IsMonitoringEnabled>
-    <ShouldDisplayAlgorithmInfo>false</ShouldDisplayAlgorithmInfo>
+    <ShouldDisplayAlgorithmInfo>false</ShouldDisplayAlgorithmInfo> <!-- change -->
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: Remove comment

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 9]
```diff
@@ -74,26 +74,44 @@
...
+        <VertexDistanceRatioCut>500.</VertexDistanceRatioCut>
+        <PathLengthRatioCut>1.012</PathLengthRatioCut>
+        <ShowerWidthRatioCut>0.2</ShowerWidthRatioCut>
+	</algorithm>
     <algorithm type = "LArCandidateVertexCreation">
         <InputClusterListNames>ClustersU ClustersV ClustersW</InputClusterListNames>
         <OutputVertexListName>CandidateVertices3D</OutputVertexListName>
         <ReplaceCurrentVertexList>true</ReplaceCurrentVertexList>
         <EnableCrossingCandidates>false</EnableCrossingCandidates>
+ 		<ReducedCandidates>true</ReducedCandidates>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: Indentation

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 10]
```diff
@@ -74,26 +74,44 @@
...
     </algorithm>
-    <algorithm type = "LArEnergyKickVertexSelection">
+	<algorithm type = "LArBdtVertexSelection">
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: Indentation

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 11]
```diff
@@ -74,26 +74,44 @@
...
+            <tool type = "LArGlobalAsymmetryFeature"/>
+            <tool type = "LArShowerAsymmetryFeature"/>
+            <tool type = "LArRPhiFeature"/>
         </FeatureTools>
-    </algorithm>
+	</algorithm>
+	<algorithm type = "LArCutClusterCharacterisation">
+        <InputClusterListNames>ClustersU ClustersV ClustersW</InputClusterListNames>
+        <ZeroMode>true</ZeroMode>
+	</algorithm>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: Indentation

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 12]
```diff
@@ -317,6 +335,37 @@
...
+        <SvmName>PfoCharacterisation</SvmName>
+        <SvmFileNameNoChargeInfo>BdtExampleNoChargeInfo_NTrees_800_TreeDepth_3.xml</SvmFileNameNoChargeInfo>
+        <SvmNameNoChargeInfo>PfoCharacterisationNoChargeInfo</SvmNameNoChargeInfo>
+        <TrainingSetMode>false</TrainingSetMode>
+        <TrainingOutputFileName>PandoraBDT_TrainingOutput_Hierarchy_ML_version_two</TrainingOutputFileName>
+        <FeatureTools>
+            <tool type = "LArThreeDLinearFitFeatureTool"/>
+            <tool type = "LArThreeDVertexDistanceFeatureTool"/>
+            <tool type = "LArThreeDPCAFeatureTool"/>
+			<tool type = "LArPfoHierarchyFeatureTool"/>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: Indentation

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 13]
```diff
@@ -317,6 +335,37 @@
...
+            <tool type = "LArThreeDOpeningAngleFeatureTool"><HitFraction>0.2</HitFraction></tool>
+            <tool type = "LArThreeDChargeFeatureTool"/>
+        </FeatureTools>
+        <FeatureToolsNoChargeInfo>
+            <tool type = "LArThreeDLinearFitFeatureTool"/>
+            <tool type = "LArThreeDVertexDistanceFeatureTool"/>
+            <tool type = "LArThreeDPCAFeatureTool"/>
+			<tool type = "LArPfoHierarchyFeatureTool"/>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: Indentation

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 14]
```diff
@@ -317,6 +335,37 @@
...
+            <tool type = "LArThreeDOpeningAngleFeatureTool"><HitFraction>0.2</HitFraction></tool>
+			<WriteToTree>true</WriteToTree>
+			<OutputTree>mousam</OutputTree>
+			<OutputFile>mousam.root</OutputFile>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: Indentation and change tree and filenames

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 15]
```diff
@@ -317,6 +335,37 @@
...
+			<tool type = "LArPfoHierarchyFeatureTool"/>
+            <tool type = "LArThreeDOpeningAngleFeatureTool"><HitFraction>0.2</HitFraction></tool>
+            <tool type = "LArThreeDChargeFeatureTool"/>
+        </FeatureTools>
+        <FeatureToolsNoChargeInfo>
+            <tool type = "LArThreeDLinearFitFeatureTool"/>
+            <tool type = "LArThreeDVertexDistanceFeatureTool"/>
+            <tool type = "LArThreeDPCAFeatureTool"/>
+			<tool type = "LArPfoHierarchyFeatureTool"/>
+            <tool type = "LArThreeDOpeningAngleFeatureTool"><HitFraction>0.2</HitFraction></tool>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: I would move the HitFraction tag to a new line.

File: settings/PandoraSettings_Neutrino_DUNEFD.xml [hunk 16]
```diff
@@ -317,6 +335,37 @@
...
+        <SvmFileNameNoChargeInfo>BdtExampleNoChargeInfo_NTrees_800_TreeDepth_3.xml</SvmFileNameNoChargeInfo>
+        <SvmNameNoChargeInfo>PfoCharacterisationNoChargeInfo</SvmNameNoChargeInfo>
+        <TrainingSetMode>false</TrainingSetMode>
+        <TrainingOutputFileName>PandoraBDT_TrainingOutput_Hierarchy_ML_version_two</TrainingOutputFileName>
+        <FeatureTools>
+            <tool type = "LArThreeDLinearFitFeatureTool"/>
+            <tool type = "LArThreeDVertexDistanceFeatureTool"/>
+            <tool type = "LArThreeDPCAFeatureTool"/>
+			<tool type = "LArPfoHierarchyFeatureTool"/>
+            <tool type = "LArThreeDOpeningAngleFeatureTool"><HitFraction>0.2</HitFraction></tool>
```
Comment (AndyChappell) [inline settings/PandoraSettings_Neutrino_DUNEFD.xml]: I would move the HitFraction tag to a new line

File: settings/PandoraSettings_Master_DUNEFD.xml (hunk 1 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Master_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Master_DUNEFD.xml (hunk 2 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Master_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Master_DUNEFD.xml (hunk 3 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Master_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Master_DUNEFD.xml (hunk 4 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Master_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Master_DUNEFD.xml (hunk 5 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Master_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Neutrino_DUNEFD.xml (hunk 12 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Neutrino_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Neutrino_DUNEFD.xml (hunk 13 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Neutrino_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Neutrino_DUNEFD.xml (hunk 14 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Neutrino_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Neutrino_DUNEFD.xml (hunk 15 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Neutrino_DUNEFD.xml]: Changed.

File: settings/PandoraSettings_Neutrino_DUNEFD.xml (hunk 16 above)
Comment (MoteyMousam) [reply settings/PandoraSettings_Neutrino_DUNEFD.xml]: Changed.

**Pull request closed without merging**